src/libclipper/include/clipper/future.hpp

```diff
@@ -0,0 +1,105 @@
+#ifndef CLIPPER_LIB_FUTURE_HPP
+#define CLIPPER_LIB_FUTURE_HPP
+
+#include <atomic>
+#include <cassert>
+#include <memory>
+#include <thread>
+#include <vector>
+
+#include "boost/thread.hpp"
+
+namespace clipper {
+
+namespace future {
+
+/*
+ * The first element in the pair is a future that will complete when
+ * all the futures provided have completed. The second element is a vector
+ * of futures that have the same values and will complete at the same time
+ * as the futures passed in as argument.
+ *
+ * \param futures A list of futures to compose
+ * \param num_completed A counter indicating the number of futures that
+ * have completed so far. The reason this is an argument to the function
+ * instead of internal state is that the caller of `when_all` must make sure
+ * that the counter stays in scope for the lifetime of all the futures.
+*/
+template <class T>
+std::pair<boost::future<void>, std::vector<boost::future<T>>> when_all(
+    std::vector<boost::shared_future<T>> futures,
+    std::shared_ptr<std::atomic<int>> num_completed) {
+  if (futures.size() == 0) {
+    return std::make_pair(boost::make_ready_future(),
+                          std::vector<boost::future<T>>{});
+  }
+  int num_futures = futures.size();
+  auto completion_promise = std::make_shared<boost::promise<void>>();
+  std::vector<boost::future<T>> wrapped_futures;
+  for (auto f = futures.begin(); f != futures.end(); ++f) {
+    wrapped_futures.push_back(f->then(
```

Comment:
So you raise a good point. It's a little confusing and the documentation isn't great. As far as I can tell, `future::then` does not create a new thread. I believe this is dependent on the default [`launch_policy`](http://www.boost.org/doc/libs/1_63_0/doc/html/thread/synchronization.html#thread.synchronization.futures.reference.unique_future.then). It appears from the documentation that the default launch policy is `launch::none` which means that the continuation is called on an unspecified thread of execution. In practice, when running this on my laptop at least, no new threads are created by the call to `future::then`. On the other hand, `when_any` and `when_all` definitely create new threads.

Here's the experiment I used to verify this, you can try running it yourself. Check out [`dcrankshaw/scheduler-prototype`](https://github.com/dcrankshaw/clipper/tree/scheduler-prototype). In `query_processor.cpp`, there are [two versions of `QueryProcessor::predict`](https://github.com/dcrankshaw/clipper/blob/34a93051d670b7c78563608bd2e9c7e47cf460ad/src/libclipper/src/query_processor.cpp#L105-L285). The uncommented one uses the Clipper implementations of the `when_all` and `when_any`. The commented out version uses the Boost implementations. Neither version specifies a launch policy for the calls to `future::then`. Included in the branch is a benchmark. You can run the benchmark on each version of the predict function and you'll see that the Boost implementation has worse performance and crashes when the thread ulimit is reached.

__Here are instructions for running the benchmark:__

Start the containers the benchmark expects (two noop containers):
```bash
cd containers/python
./start_containers
```

Build and run the benchmark:
```bash
./configure --release && cd release
make bench
./src/frontends/bench
# It will prompt you for the number of requests to make. Try 100000.
```

When you're done with the experiment, Ctrl-C the containers script and it will stop the container processes for you.